crates/dbsp/src/circuit/dbsp_handle.rs

```diff
@@ -1286,29 +1261,21 @@ impl DBSPHandle {
         Ok(())
     }
 
-    fn checkpointer(&mut self) -> Result<&mut Checkpointer, DbspError> {
+    fn checkpointer(&self) -> Result<&Arc<Mutex<Checkpointer>>, DbspError> {
         self.checkpointer
-            .as_mut()
+            .as_ref()
             .ok_or(DbspError::Storage(StorageError::StorageDisabled))
     }
 
-    fn checkpoint_as(
-        &mut self,
-        uuid: Uuid,
-        identifier: Option<String>,
-        steps: Option<u64>,
-        processed_records: Option<u64>,
-    ) -> Result<CheckpointMetadata, DbspError> {
-        let checkpoint_dir = Checkpointer::checkpoint_dir(uuid);
-        self.broadcast_command(Command::Checkpoint(checkpoint_dir), |_, _| {})?;
-        self.checkpointer()
-            .unwrap()
-            .commit(uuid, identifier, steps, processed_records)
+    /// Allows creating a new checkpoint by taking consistent snapshot of the
+    /// state in dbsp.
+    pub fn checkpoint(&mut self) -> CheckpointBuilder<'_> {
+        CheckpointBuilder::new(self)
     }
 
     /// List all currently available checkpoints.
     pub fn list_checkpoints(&mut self) -> Result<Vec<CheckpointMetadata>, DbspError> {
-        self.checkpointer()?.list_checkpoints()
+        self.checkpointer()?.lock().unwrap().list_checkpoints()
```

Comment:
unwrap?
why not `?`?

Reply:
We don't usually try to propagate lock poisoning errors upward.